### PR TITLE
generic fst supports constfst and vectorfst (without FstHolder)

### PR DESCRIPTION
**Pull request by LvHang**
This PR goes on #1548 
This version has removed the FstHolder. That means only "if" branch in *-latgen-faster is fixed in order to speeding up and saving memory about decoding.

@danpovey 
The test about nnet3 tdnn training has passed with swbd corpus.
Now, I'm testing the whole pipeline(run.sh) with mini_librispeech corpus. I will notice you when it accomplish.
Bests,
Hang

### CODE REVIEW

**Comment by danpovey** [comment]
Thanks!  I'll check it when you say the mini-librispeech tests are done.

**Comment by LvHang** [comment]
Hi Dan,
I think the mini-librispeech test is done. 
The specific results are in /export/a11/hlyu/kaldi-20170531/egs/mini_librispeech/s5.
Maybe you can review the codes. Thanks!
Hang

File: src/fstext/kaldi-fst-io.h [hunk 1]
```diff
@@ -41,6 +41,22 @@ namespace fst {
...
+// (const-fst can give better performance for decoding).
+Fst<StdArc> *ReadFstKaldiGeneric(std::string rxfilename,
+                                 bool throw_on_err = true);
+
+// This function attempts to dynamic_cast the pointer 'fst' (which will likely
+// have been returned by ReadFstGeneric()), to the more derived 
+// type VectorFst<StdArc>. If this succeeds, it returns the same pointer;
+// if it fails, it converts the FST type (by creating a new VectorFst<stdArc>
+// initialized by 'fst'), prints a warning, and deletes 'fst'.
+VectorFst<StdArc> *CastToVectorFst(Fst<StdArc> *fst);
```
Comment (danpovey) [inline src/fstext/kaldi-fst-io.h]: Actually, let's rename this function to CastOrConvertToVectorFst(), which is more informative.

Comment (danpovey) [inline src/fstext/kaldi-fst-io.h]: No need to re-est after that change, just make sure it compiles.

**Comment by LvHang** [comment]
I have already renamed the function and it can be compiled.

**Comment by danpovey** [comment]
Thanks, I'll probably merge once the automated checks complete.

**Comment by danpovey** [comment]
Thanks!  merged.

**Pull request merged**